Add ZSTD decompression support

IoplachkinI opened the pull request:
Resolves issue #549

Benchmark results:
```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
ZstdDecompress/1024        4321 ns         4329 ns       161878
ZstdDecompress/2048        5288 ns         5293 ns       128786
ZstdDecompress/4096        7116 ns         7122 ns        97136
ZstdDecompress/8192       11499 ns        11502 ns        59946
ZstdDecompress/16384      16596 ns        16595 ns        42723
ZstdDecompress/32768      26929 ns        26926 ns        26055
GzipDecompress/1024       10459 ns        10457 ns        66482
GzipDecompress/2048       16552 ns        16550 ns        42502
GzipDecompress/4096       29337 ns        29338 ns        23916
GzipDecompress/8192       57317 ns        57314 ns        12272
GzipDecompress/16384     113640 ns       113633 ns         6156
GzipDecompress/32768     230311 ns       230269 ns         2952
```

VS-CDR committed:
implement zstd decompress stream wrapper

VS-CDR committed:
add check for max_sz

VS-CDR committed:
fix indents

VS-CDR committed:
optimized output creating

IoplachkinI committed:
Add error name to error msg

IoplachkinI committed:
Update Decompress()

IoplachkinI committed:
Add working tests

IoplachkinI committed:
Add basic zstd support in DecompressRequestBody()

IoplachkinI committed:
Add tests, integration tests and benchmarks

IoplachkinI committed:
Fix random character generator

IoplachkinI committed:
Fix py test

IoplachkinI committed:
Fix leaks in tests

IoplachkinI committed:
Remove unnecessary include

VS-CDR committed:
reduce bufsz

VS-CDR committed:
move buffer from stack to dram

IoplachkinI committed:
Fix tests and benches

IoplachkinI committed:
Organize file structure

IoplachkinI committed:
Reset core/CMakeLists.txt

IoplachkinI committed:
Remove const cast

apolukhin commented on universal/src/compression/zstd.cpp:
```diff
@@ -0,0 +1,58 @@
+    cur_pos += input.size;
+  }
+
+  ZSTD_freeDStream(stream);
```
Use RAII for that to avoid resource leaks

VS-CDR replied:
Done

VS-CDR committed:
add temporary needed free

VS-CDR committed:
add RAII

VS-CDR committed:
hotfix

VS-CDR committed:
hotfix

VS-CDR committed:
small refactoring

VS-CDR committed:
hotfix

VS-CDR committed:
small refactoring

VS-CDR committed:
fix indents

VS-CDR committed:
attempt to fix

'В случае если content_encoding != "gzip" и content_encoding != "zstd" тело запроса потрётся :('

VS-CDR committed:
added ZSTD_getFrameContentSize

botanegg commented on universal/src/compression/zstd.cpp:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  auto stream_del = [](ZSTD_DStream* ptr) { ZSTD_freeDStream(ptr); };
+  auto stream_guard =
```
seems like you can move up this guard to line after ZSTD_createDStream
also ZSTD_freeDStream accept NULL pointer

botanegg commented on universal/src/compression/zstd.cpp:
```diff
@@ -0,0 +1,95 @@
+        throw TooBigError();
+      }
+      if (decompressed_size > std::numeric_limits<int64_t>::max() / 2) {
+        return DecompressStream(compressed, max_size);
```
Are you sure you want to decompress more than 4611686018427387903 bytes?
seems like a sign mismatch here for `<>`
also decompressed_size is unsigned long long

VS-CDR committed:
corrected some remarks

VS-CDR committed:
extra refactoring

VS-CDR committed:
hotfix

VS-CDR committed:
add const qualifier

VS-CDR committed:
add const qualifier

VS-CDR committed:
add const qualifier

VS-CDR committed:
Merge branch 'userver-framework:develop' into add-zstd-support

VS-CDR committed:
little refactoring and adding test on check max_sz for zstd

VS-CDR committed:
hotfix gzip method!!!

VS-CDR committed:
tests rework and hotfix in stream method

VS-CDR committed:
code refactoring

VS-CDR committed:
name collision fix

VS-CDR committed:
add const qualifier

VS-CDR committed:
...
